app/controllers/invites_controller.rb

```diff
@@ -40,24 +41,29 @@ def show
 
   def perform_accept_invitation
     params.require(:id)
-    params.permit(:username, :name, :password, :timezone, user_custom_fields: {})
+    params.permit(:email, :username, :name, :password, :timezone, user_custom_fields: {})
     invite = Invite.find_by(invite_key: params[:id])
 
     if invite.present?
       begin
-        user = invite.redeem(username: params[:username], name: params[:name], password: params[:password], user_custom_fields: params[:user_custom_fields], ip_address: request.remote_ip)
+        user = if invite.is_invite_link?
+          invite.redeem_invite_link(email: params[:email], username: params[:username], name: params[:name], password: params[:password], user_custom_fields: params[:user_custom_fields], ip_address: request.remote_ip)
+        else
+          invite.redeem(username: params[:username], name: params[:name], password: params[:password], user_custom_fields: params[:user_custom_fields], ip_address: request.remote_ip)
+        end
+
         if user.present?
           log_on_user(user) if user.active?
           user.update_timezone_if_missing(params[:timezone])
           post_process_invite(user)
+          response = { success: true, message: I18n.t('invite.confirm_email') }
+        else
+          response = { success: false, message: I18n.t('invite.not_found_json') }
         end
 
-        response = { success: true }
         if user.present? && user.active?
           topic = invite.topics.first
           response[:redirect_to] = topic.present? ? path("#{topic.relative_url}") : path("/")
-        else
-          response[:message] = I18n.t('invite.confirm_email')
```

Comment:
Just confirming that this is no longer required? Can the `user` still be `present?` but not `active?`

Reply:
Yes, that can be the case, that is why I moved this block to `user.present?` check just above, as per: https://github.com/discourse/discourse/pull/9813/files#diff-322a08e87c3cc46449eac6235452f266R59

Reply:
I think something is still off here. If `user.present?` is true, we'll always set `message: I18n.t('invite.confirm_email')`. However, if a `user.active?` the message shouldn't be set. You can see this in the payload when running the following spec `describe "directs to the first topic the user was invited to"`

Reply:
Good catch. I've fixed this case.